### PR TITLE
Extract ItemReconciler service

### PR DESCRIPTION
**Pull request by lucaswerkmeister**
This service wraps a `ReconciliationService` and `SimplePutStrategy` and extracts the reconciliation URL from an input item before passing it to the `ReconciliationService` and `SimplePutStrategy`. Its return value is an Item and a revision ID, but the meaning is quite different, so we introduce a new result class, `ReconciledItem`, rather than reusing `ReconciliationItem`.

Bug: [T283567](https://phabricator.wikimedia.org/T283567)

---

Includes one preparation commit to move some classes.

The names “ItemReconciler” and “ReconciledItem” aren’t that great… any other ideas?

### CODE REVIEW

File: includes/MediaWiki/Api/EditEndpoint.php [hunk 1]
```diff
@@ -153,14 +111,14 @@ public function run() {
...
-			$reconciliationItem->getRevision()
+			$reconciledItem->getBaseRevisionId()
 		);
 
 		if ( $saveStatus->isOK() ) {
 			$saveStatus->merge( $editEntity->attemptSave(
 				$toSave,
 				'Reconciliation Edit',
-				$reconciliationItem->getRevision() ? null : EDIT_NEW,
+				$reconciledItem->isNew() ? EDIT_NEW : 0,
```
Comment (toban) [inline includes/MediaWiki/Api/EditEndpoint.php]: Shouldn't this be `EDIT_UPDATE`?

Seems like the 0 or null becomes a `EDIT_AUTOSUMMARY` by default? Not sure if it even matters.

File: includes/Reconciliation/ReconciliationItem.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * This class contains an item as it was found in the database,
+ * or a newly created item if no existing item was found,
+ * in which case the revision ID is false.
+ * It should not be confused with {@link ReconciledItem},
+ * which includes modifications made to the item based on the input.
+ *
+ * @license GPL-2.0-or-later
+ */
+class ReconciliationItem {
```
Comment (toban) [inline includes/Reconciliation/ReconciliationItem.php]: How do you feel about naming this `ReconciliationServiceItem` ?

Even though we have the comment not to confuse the two i think making the difference a bit more obvious will make this less likely to be confused?

**Review by toban — APPROVED** [review]
Looks good to me, just a few thoughts that popped up.

File: includes/MediaWiki/Api/EditEndpoint.php (hunk 1 above)
Comment (lucaswerkmeister) [reply includes/MediaWiki/Api/EditEndpoint.php]: Hm, `EDIT_UPDATE` might make sense. But where do you see `EDIT_AUTOSUMMARY` coming in?

File: includes/Reconciliation/ReconciliationItem.php (hunk 2 above)
Comment (lucaswerkmeister) [reply includes/Reconciliation/ReconciliationItem.php]: Sounds okay to me. Or maybe even `ReconciliationServiceResult`, without “item”?

File: includes/MediaWiki/Api/EditEndpoint.php (hunk 1 above)
Comment (toban) [reply includes/MediaWiki/Api/EditEndpoint.php]: I thought it was set here https://github.com/wikimedia/Wikibase/blob/REL1_35/repo/includes/EditEntity/MediawikiEditEntity.php#L739 if left out

Comment (lucaswerkmeister) [reply includes/MediaWiki/Api/EditEndpoint.php]: Okay, but it doesn’t look like we can influence this, so I’d leave it out of this call.

File: includes/Reconciliation/ReconciliationItem.php (hunk 2 above)
Comment (lucaswerkmeister) [reply includes/Reconciliation/ReconciliationItem.php]: Okay, I’ll insert another commit to rename it to `ReconciliationServiceResult`.

Comment (lucaswerkmeister) [reply includes/Reconciliation/ReconciliationItem.php]: Hm, on the other hand the term “item” is used a lot for these results (e.g. `EditRequest::otherItems()`), and changing those to “result” would probably make it less clear overall.

Let’s go with `ReconciliationServiceItem`.

**Pull request merged**